### DISCUSSION TITLE
PHP library for Greenfield API

### DISCUSSION THREAD

**Discussion by ndeet** [btcpayserver/btcpayserver#2663] [Plugins & integrations]
As a sub-discussion of https://github.com/btcpayserver/btcpayserver/discussions/2661 where we can discuss the requirements for the PHP library we want to build as a base for all PHP related integrations. 

(I will update this post according to our discussion outcomes)

### Autogenerated Code vs Custom library

It seems that while autogenerated code from the swagger spec lowers the maintenance burden the developer experience (DX) is not that great and works a bit against having a easy to use base library for all related PHP integrations.

### Library considerations

- Composer package as first class citizen
- but for legacy data structure projects it should also support manual includes or provide a downloadable zip where you can just include `/vendor/autoload.php` 
- avoid usage of 3rd party libraries like Symfony or Guzzle (HTTP client) and instead use widely supported PHP core extensions like cURL
- provide specific Exception classes for every API error that can happen

### Class hierarchy / data structure

Like mentioned above we want a clean and standalone solution without any 3rd party dependency. Also the class structure should be kept KISS and avoid overengineering (e.g. like having a `User` base class and a basically empty extending `Customer` class; or have interfaces everywhere without any use of it

Please let post some of your favourite client libraries (can also be different languages than PHP) here so we can decide on what looks like a good DX and flexible enough structure.

I looked into Stripes PHP lib and while I do not think we need all the bells and whistles they have a nice approach of having their own [cURL client](https://github.com/stripe/stripe-php/blob/master/lib/HttpClient/CurlClient.php) (and not depending on Guzzle) and also a [test-boilerplate](https://github.com/stripe/stripe-php/tree/master/tests) how we can mock Greenfield API server using the swagger/openapi spec to run tests. 

### Scope for MVP

The first target is imo to have feature parity with the BitPay client in the sense that we support all the basic operations so we can start building new versions of e.g. the WooCommerce plugin. We do not need all that complicated auth logic and signing anymore but just support the new API token (I assume). So this means support for:
- authorization via API key (is there any need to support username + password aka basic auth for now?)
- create invoice
- fetch existing invoice data
- currency
- (what else?)

Feel free to add your ideas and let's discuss.

**Comment by woutersamaey** [comment]
I also want Exception classes for each and every API exception there is so we can have proper catching and error handling!

**Reply by ndeet** [reply]
Very good point 👍

**Comment by woutersamaey** [comment]
I'm torn.

Using our own OOP-version of cURL may be a bit re-inventing the wheel.
On the other hand there is PSR-18 (https://www.php-fig.org/psr/psr-18/) that's perfect for us.... but we don't want extra dependancies.

Should we copy-paste / bundle a lightweight PSR18 HTTP client without composer?

Theory says no.
Practice says yes?

**Reply by ndeet** [reply]
> Should we copy-paste / bundle a lightweight PSR18 HTTP client without composer?
> 
> Theory says no.
> Practice says yes?

I think as long as it is something like a single file HTTP-client I would say, yes could work. Copying a whole library I'm really struggling to justify it not using just as a composer dependency (and make sure we also provide a downloadable package with all the deps if someone is not using Composer). If we decide to do that then why not go with Guzzle [1] then. Depending on a HTTP library might be fine anyway - on the other hand I like how Stripe is usable standalone without any other dependency and not make you run into dependency if there is a major HTTP client version mismatch between the library and the consuming project (e.g. Magento or Drupal etc.)

I did some quick search but did not find a lightweight, ideally single file, PSR-18 HTTP-client. Do you have any in mind to for me to check?

[1] I can remember there was major dependency hell with some Guzzle upgrades (afair 5 to 6) which then led to PSR-18 from what I read in the blogpost. So likely this should not happen again but what we need to decide for us is if we go with a HTTP-client library dependency or not I guess.

**Reply by woutersamaey** [reply]
I agree on the single file part. Let's search on it and if it does not exist, I'm willing to code one.

**Reply by woutersamaey** [reply]
As it turns out, using PSR-18 forces us to include the PSR18 interfaces and that happens using... composer...!
So, let's just keep it simple and program our own cURL calls.
This is getting too complex I believe.

**Reply by ndeet** [reply]
Yes I looked into it a bit. When you go PSR-18 you also go PSR-7 and then the solution seems to be httplug which adds some interfaces and they you are implementation independent but still need to provide a client (or use an existing one) of course all using Composer. On one hand should not be a problem if we do a build script running composer install and providing a package with /vendor inside for use for legacy projects not using Composer. On the other hand we will have quite some dependencies in there even if only lightweight and it feels a bit overengineered. 

I'm a bit undecided here, benefit would be to not reinventing the wheel and to have an abstraction to the actual client and downstream projects can plug in their own - but on the other hand providing our own slim client seems much more straight forward and keeping dependencies low.

I think this is why Stripe and others went with their own small cURL client and some Request/Response classes and not going PSR-18/7, I checked a few libraries and did not yet find one going PSR-18/7 .e.g. see also [Square PHP SDK](https://github.com/square/square-php-sdk/tree/master/src/Http), PayPal and others are similar. 

To check what others do [Facebook SDK](https://github.com/facebookarchive/php-graph-sdk/tree/5.x/src/Facebook/HttpClients) uses Guzzle and provides also cURL a client. It looks similar to Stripe but more clean and providing more Client implementations. We might can use that as an inspiration and for now only go with cURL for now. I will look into this a bit deeper.

**Reply by ndeet** [reply]
Re Facebook SDK: seems while 5.5 is still stable they go for httplug (PSR-18/7) with their v6: https://github.com/facebookarchive/php-graph-sdk/tree/master/

I guess I will start some small PoC with both approaches to see what makes more sense.

**Reply by woutersamaey** [reply]
@ndeet this is getting too complex I believe.

What we need is simple, not worth the special builds etc.
You can argue that we re-invent the wheel, but it's certainly less than 50-100 lines of code... so what?
The build system you're describing is more complex than this.
Others may also struggle to integrate if they already have composer / autoloaders.

Nobody is going to want a PSR18 compatible client because it matters to them? It doesn't.

We just need:
- A HTTP Client class. Can even use a static method to do the request like: `public static doRequest(string $method, string $url, string|array $body, array $headers)`
- A response class with body, status and headers fields, read-only
- Exception classes

That's it.

- We don't need to handle cookies
- We don't need a request class
- We don't need async stuff
- We don't need parallel requests

**Reply by ndeet** [reply]
Agree, so like I suggested in the description we might just do it similar like Stripe with their own lightweight client, also Square does it like this. I would still do an Interface (with only doRequest()/send()) like they do so that one can provide their own client if needed. 

I hoped there is a best practice/boilerplate that everyone uses nowadays like PSR-18 so we do not add another custom client but like you said probably not worth it spending all the time for this at the beginning. We can revisit next major version I guess. 

So let's decide to go with a small lightweight cURL client for now 👍

**Reply by woutersamaey** [reply]
Do we have a repo where we can start building?

**Reply by ndeet** [reply]
Currently there is only the old libs and the one generated from openapi/swagger here https://github.com/btcpayserver/btcpayserver-greenfield-php (not even sure if anybody tested if it works at all)

We might want to delete or force-push that at some point when we come up with the custom greenfield client. Or we start a new one with btcpayserver-greenfield-php-client or similar in the meantime. I will start something tonight or tomorrow at the latest, we can do on our own profiles in the meantime I guess. Will let you know as soon as I have something put together.

**Reply by ndeet** [reply]
@Kukks @pavlenex any opinion if we should keep https://github.com/btcpayserver/btcpayserver-greenfield-php for now and start a new one in parallel like suggested above or should we force-push/overwrite it at some point?

**Reply by Kukks** [reply]
We can nuke it with no worries.

**Reply by woutersamaey** [reply]
@ndeet I have already implemented 2 calls with a couple of classes around them (needed this for something else I'm working on that needs to move fast), so if you want, I can share and build on this. It's very easy and straightforward.

**Reply by ndeet** [reply]
@woutersamaey would be great, I also started and will have something to share in the next couple of hours

**Reply by woutersamaey** [reply]
@ndeet or @pavlenex can you provide me access to a repo or do I publish under my own name (for now)?

**Reply by ndeet** [reply]
@woutersamaey yes, I would say until we have something more complete let's do on personal profile and then overwrite above mentioned https://github.com/btcpayserver/btcpayserver-greenfield-php repo. We already have 2 other PHP library repos and on official GH we should imo not create too many just for testing/trying things out.

**Reply by woutersamaey** [reply]
@ndeet could we maybe schedule a (short) phone call? I'll publish my code and we can go through it. I'm in Belgium (UTC +2) and free most evenings after 20:00.

**Reply by ndeet** [reply]
@woutersamaey I'm also in UTC+2, Austria :) Mostly available during working hours but I can do around 21:00.

**Comment by woutersamaey** [comment]
Work in progress is at https://github.com/woutersamaey/btcpayserver-greenfield-php-client for now.
